src/VisualStudio/Core/Def/ValueTracking/ValueTrackingCommandHandler.cs

```diff
@@ -154,43 +159,57 @@ private async Task ShowToolWindowAsync(ITextView textView, ISymbol selectedSymbo
             var document = solution.GetRequiredDocument(location.SourceTree);
 
             var sourceText = await location.SourceTree.GetTextAsync(cancellationToken).ConfigureAwait(false);
-            sourceText.GetLineAndOffset(location.SourceSpan.Start, out var lineStart, out var _);
-            sourceText.GetLineAndOffset(location.SourceSpan.End, out var lineEnd, out var _);
-            var lineSpan = LineSpan.FromBounds(lineStart, lineEnd);
-
             var documentSpan = await ClassifiedSpansAndHighlightSpanFactory.GetClassifiedDocumentSpanAsync(document, location.SourceSpan, cancellationToken).ConfigureAwait(false);
             var classificationResult = await ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync(documentSpan, cancellationToken).ConfigureAwait(false);
 
             var root = new ValueTrackingTreeItemViewModel(
                 document,
-                lineSpan.Start,
+                location.SourceSpan,
                 sourceText,
                 selectedSymbol,
                 classificationResult.ClassifiedSpans,
-                classificationFormatMap,
-                _typeMap,
+                toolWindow.ViewModel,
                 _glyphService,
-                childViewModels);
+                _threadingContext,
+                children: childViewModels);
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-            await ShowToolWindowAsync(root, cancellationToken).ConfigureAwait(false);
+            toolWindow.ViewModel.Roots.Clear();
+            toolWindow.ViewModel.Roots.Add(root);
+
+            await ShowToolWindowAsync(cancellationToken).ConfigureAwait(true);
 
             ValueTrackingTreeItemViewModel CreateViewModel(ValueTrackedItem valueTrackedItem, ImmutableArray<ValueTrackingTreeItemViewModel> children = default)
                 => new ValueTrackedTreeItemViewModel(
                    valueTrackedItem,
                    solution,
-                   classificationFormatMap,
-                   _typeMap,
+                   toolWindow.ViewModel,
                    _glyphService,
                    valueTrackingService,
+                   _threadingContext,
                    children);
         }
 
-        private async Task ShowToolWindowAsync(ValueTrackingTreeItemViewModel viewModel, CancellationToken cancellationToken)
+        private async Task ShowToolWindowAsync(CancellationToken cancellationToken)
+        {
+            var roslynPackage = await TryGetRoslynPackageAsync(cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(roslynPackage);
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            await roslynPackage.ShowToolWindowAsync(
```

Comment:
it's a pity ShowToolWindow doesn't do teh switch, given that's async.  @sharwell should we invert this?

Reply:
It may, but I haven't found good documentation on this. AFAIK switchtomain should no-op if it's already on main so the work is done regardless of who does it

Reply:
[internal link](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fshell%2FPackageFramework%2FCurrent%2FShell%2FAsyncPackage.cs&version=GBmain&line=1045&lineEnd=1046&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) looks like this already switched to the main thread internally.